File: src/pyodide/nblast.py
"""NBLAST, as Coda asks for it.

This file is the whole of the Python side. It is loaded as a string (vite's `?raw`) and
executed once per runtime boot, so everything here is a definition — the worker calls in
through `coda_dotprops` and `coda_nblast`.

**It is a real `.py` rather than a template literal in the worker** so that it can be read
with syntax highlighting, diffed, and — the half that matters — *run directly against the
same wheel*. Every claim in the comments below was established that way rather than
recalled.

Points arrive in **micrometres**. That is not a preference: the FCWB scoring matrix
fastcore embeds has its last distance bin at 40 um, and beyond it every cell is about -10.
Handed nanometres, every pair of neurons in a dataset scores like two neurons that have
never met — a confident, uniform, entirely wrong answer with no error anywhere. The
conversion happens in `nblastOps.ts`, on the JS side, where the skeletons still know they
are in nm.
"""

import numpy as np
import navis_fastcore as fc


def _coda_prepare(req, k, resample, report=None):
    """Both sides of a comparison as fastcore `Dotprop`s.

    The step every entry point here starts with, and the place the progress split lives: a
    query-only call spends more of the bar on its one set than a query/target call spends on
    its first. Shared so the two entry points cannot come to disagree about either.
    """
    target_set = req.get("target")

    def phase(lo, hi):
        if report is None:
            return None
        return lambda f: report(lo + (hi - lo) * f, "tangent vectors")

    split = 0.25 if target_set else 0.45
    query = coda_dotprops(req["query"], k, resample, phase(0.0, split))
    target = coda_dotprops(target_set, k, resample, phase(split, 0.5)) if target_set else None
    return query, target


def coda_nblast_run(request, report=None):
    """The whole of one NBLAST: resample, tangent vectors, score.

    Composed here rather than in TypeScript so the bridge stays one call. Progress is reported
    across the two phases this can see — dotprops, then the blast — and the blast itself is
    opaque for a measured reason, recorded on `coda_nblast` below.
    """
    req = request.to_py()
    query, target = _coda_prepare(req, int(req["k"]), float(req["resample"]), report)

    if report is not None:
        rows, cols = len(query), len(target) if target else len(query)
        report(0.5, f"scoring {rows} x {cols}")

    scores = coda_nblast(
        query, target, bool(req["normalize"]), str(req["symmetry"]), bool(req["useAlpha"])
    )
    return {
        "scores": scores.ravel(),
        "rows": scores.shape[0],
        "cols": scores.shape[1],
    }


def coda_nblast_knn_run(request, report=None):
    """The k nearest neighbours of each neuron, without the full matrix.

    The same three steps as `coda_nblast_run` up to the scoring, and then a different
    question: fastcore shortlists candidates from a coarse voxel signature and scores only
    those, so the cost is `n * n_candidates` rather than `n^2`. Every score returned is an
    exact NBLAST value — only *which* pairs made the shortlist is approximate. Measured by
    fastcore on 163,976 neurons: recall@20 of 0.990 at the default 200 candidates, having
    scored 0.16% of the pairs.

    Returns the two `(n_query, k)` arrays flat, with their shape beside them. `idx` is cast to
    **int32** on the way out, deliberately: it is int64 in fastcore, and an int64 numpy array
    converts to a `BigInt64Array` in JavaScript, which nothing on the other side is expecting
    and which does not compare with a number. Neuron *counts* here are bounded by the node's
    Max neurons long before int32 is.

    Rows with fewer than `k` candidates are padded — `-1` in `idx`, `-inf` in `scores` — and
    the padding is left in place for the caller to drop, since dropping it here would lose the
    shape that makes the two arrays readable.
    """
    req = request.to_py()
    k = int(req["k"])
    query, target = _coda_prepare(req, int(req["tangentK"]), float(req["resample"]), report)

    if report is not None:
        report(0.5, f"{k} nearest of {len(target) if target else len(query)}")

    symmetry = req["symmetry"]
    idx, scores = fc.nblast_knn(
        query,
        target=target,
        k=k,
        symmetry=None if symmetry == "none" else symmetry,
        n_candidates=int(req["nCandidates"]),
        normalize=bool(req["normalize"]),
        use_alpha=bool(req["useAlpha"]),
    )
    idx = np.asarray(idx)
    scores = np.asarray(scores)
    return {
        "idx": np.ascontiguousarray(idx, dtype=np.int32).ravel(),
        "scores": np.ascontiguousarray(scores, dtype=np.float64).ravel(),
        "rows": int(idx.shape[0]),
        "k": int(idx.shape[1]),
    }


def coda_dotprops(point_set, k, resample, report=None):
    """Concatenated skeleton points in, a list of fastcore `Dotprop`s out.

    `points` is xyz interleaved, float32, every neuron laid after the last; `offsets` says
    where each one starts, counted in points rather than floats. `parents` is one parent
    index per point (-1 for a root), neuron-local — which is exactly the shape
    `resample_skeleton` wants once the node ids are the row numbers.

    **Nothing here guards the small cases, and that is deliberate.** Checked against this
    wheel: fastcore clamps `k` to the point count itself, resamples a forest without
    complaint (both roots survive), and accepts a one-point neuron. A guard here would be
    dead code asserting the opposite of what the library does.
    """
    xyz = np.frombuffer(point_set["points"], dtype=np.float32).reshape(-1, 3).astype(np.float64)
    par = np.frombuffer(point_set["parents"], dtype=np.int32)
    off = np.frombuffer(point_set["offsets"], dtype=np.int32)
    count = len(off) - 1

    out = []
    for i in range(count):
        a, b = int(off[i]), int(off[i + 1])
        pts = np.ascontiguousarray(xyz[a:b])
        if resample > 0 and b - a > 1:
            ids = np.arange(b - a, dtype=np.int32)
            # resample_skeleton returns (ids, parents, xyz, source, alpha, node_map); the
            # geometry is all NBLAST wants, and re-rooting or re-parenting is not our business.
            pts = np.ascontiguousarray(
                fc.resample_skeleton(ids, par[a:b], pts, resample)[2], dtype=np.float64
            )
        vect, alpha = fc.dotprops(pts, k=k)
        out.append(fc.Dotprop(pts, vect, alpha))
        if report is not None and (i % 8 == 7 or i == count - 1):
            report((i + 1) / count)
    return out


def coda_nblast(query, target, normalize, symmetry, use_alpha):
    """Score one set against another, or a set against itself.

    **The square case goes through `nblast_allbyall` as a single call**, which is why no
    progress is reported from inside the blast. Chunking the rows to drive a progress bar
    was measured on this wheel at 1.8x (50-row chunks) to 5.1x (10-row) the run it would be
    reporting on, for byte-identical scores. A bar that costs several times the wait it
    describes is not a bar.

    `symmetry=None` is fastcore's own default and means "no symmetry"; Coda passes `mean`
    unless asked otherwise, because a heatmap that disagrees with itself across the diagonal
    is a puzzle rather than a finding.
    """
    sym = None if symmetry == "none" else symmetry
    if target is None:
        scores = fc.nblast_allbyall(
            query, normalize=normalize, symmetry=sym, use_alpha=use_alpha
        )
    else:
        scores = fc.nblast(
            query, target, normalize=normalize, symmetry=sym, use_alpha=use_alpha
        )
    # C-contiguous float64 is what `getBuffer('f64')` on the JS side reads without a copy.
    return np.ascontiguousarray(scores, dtype=np.float64)


def coda_synapse_sets(point_set):
    """Concatenated synapse points in, a list of `(N, 4)` arrays out.

    The syNBLAST counterpart of `coda_dotprops`, and much the smaller of the two: a synapse
    has no tangent vector to fit, so there is no `k`, no neighbourhood and no resampling —
    just the points, grouped, with a connector type in the fourth column.

    `points` is xyz interleaved, float32, every neuron laid after the last; `offsets` says
    where each one starts, counted in points; `types` is one small integer per point. The
    type column is what `by_type` compares on, and fastcore wants it numeric — the mapping
    from Coda's `polarity` strings onto 0 and 1 happens on the JavaScript side, where the
    column and its dtype are still in view.

    **Every group here is non-empty by construction.** The caller groups a point cloud by
    `neuronId`, so a neuron with no synapses has no group rather than an empty one — which
    matters because a `(0, 4)` array is a neuron fastcore has nothing to compare.
    """
    xyz = np.frombuffer(point_set["points"], dtype=np.float32).reshape(-1, 3)
    types = np.frombuffer(point_set["types"], dtype=np.int32)
    off = np.frombuffer(point_set["offsets"], dtype=np.int32)

    # One `(N, 4)` allocation for the whole cloud, with the float32 -> float64 widening
    # happening on the way *into* it. The obvious spelling — `.astype(np.float64)` first, then
    # a per-neuron `np.empty` copied into — allocates the full cloud twice and writes it twice,
    # which at the half-million synapses `checkSynblastSize` warns at is 28 MB where 16 would
    # do. The per-neuron blocks are then row slices of a C-contiguous array, which are
    # themselves C-contiguous, so fastcore takes them as views rather than as copies.
    blocks = np.empty((len(xyz), 4), dtype=np.float64)
    blocks[:, :3] = xyz
    blocks[:, 3] = types
    return [blocks[int(off[i]) : int(off[i + 1])] for i in range(len(off) - 1)]


def coda_synblast_run(request, report=None):
    """syNBLAST: compare neurons by where their synapses are rather than by their shape.

    For every query connector the nearest target connector *of the same type* is found and
    the euclidean distance scored through the same FCWB lookup matrix `coda_nblast` uses,
    with the dot product fixed at 1 — a synapse has no direction. So the scores are on the
    same scale as an NBLAST and mean something quite different: two neurons with identical
    arbors that talk to different partners in different places score low here and high there.

    **Points are micrometres**, for the reason the module docstring gives: the same lookup
    matrix runs out at a 40 um distance bin, so a set handed over in nanometres scores every
    pair as strangers, uniformly, with nothing anywhere to say why.

    No progress is reported from inside the comparison. fastcore builds an index and then
    scores, neither of which exposes a per-neuron hook, and `coda_nblast`'s measurement of
    what chunking costs applies here unchanged.
    """
    req = request.to_py()
    query = coda_synapse_sets(req["query"])
    target_set = req.get("target")
    target = coda_synapse_sets(target_set) if target_set else None

    if report is not None:
        rows, cols = len(query), len(target) if target else len(query)
        report(0.2, f"scoring {rows} x {cols}")

    symmetry = str(req["symmetry"])
    scores = fc.synblast(
        query,
        target,
        by_type=bool(req["byType"]),
        normalize=bool(req["normalize"]),
        symmetry=None if symmetry == "none" else symmetry,
    )
    scores = np.ascontiguousarray(scores, dtype=np.float64)
    return {
        "scores": scores.ravel(),
        "rows": scores.shape[0],
        "cols": scores.shape[1],
    }

File: src/pyodide/nblast.ts
/**
 * NBLAST's typed wrapper over the Python bridge.
 *
 * **This file is the shape a second capability copies.** A capability is three things and no
 * more: a `.py` registered in `runtime.ts`'s `MODULES`, the types its request and result take,
 * and a function that calls `callPython` and reads the answer by name. Nothing in `engine.ts`,
 * `worker.ts` or `types.ts` changes when one is added — that was the whole point of making the
 * protocol about calling a function rather than about scoring neurons.
 *
 * The types live here rather than in `types.ts` for the same reason: `types.ts` is the wire,
 * and what travels on it is nobody else's business.
 */

import { callPython } from './engine'
import type { CallOptions } from './engine'
import { float64From, int32From, numberFrom } from './types'

/** How a query set is scored against a target that is the same set. */
export type NblastSymmetry = 'none' | 'min' | 'max' | 'mean'

/**
 * One neuron set, flattened.
 *
 * Three buffers rather than an array of objects: a hundred skeletons is a hundred thousand
 * points, and an array of `{x, y, z}` would be a hundred thousand objects to clone at the
 * postMessage boundary. Flattened, the whole set moves as three transfers and costs nothing.
 *
 * **Points are micrometres**, converted on the way in. See `nblast.py` for why that is not a
 * preference — the embedded FCWB matrix stops at 40 um, so nanometres score every pair as
 * strangers with no error anywhere.
 */
export type PointSet = {
  /** xyz interleaved, one neuron after the last. */
  points: Float32Array
  /** Parent index per point, `-1` for a root. Neuron-local, so index 0 is that neuron's first. */
  parents: Int32Array
  /** Where each neuron starts, counted in points. Length is `count + 1`. */
  offsets: Int32Array
}

/*
 * A `type` rather than an `interface`, and the same for `PointSet` above — not a style choice.
 * TypeScript gives a type alias an implicit index signature and an interface none, so an
 * interface is not assignable to `PyArg`'s object branch and a request declared as one fails
 * to compile at the `callPython` call with a message about `undefined`. Worth knowing before
 * writing the second capability, since it is not obvious from the error.
 */
export type NblastRequest = {
  query: PointSet
  /** Absent means all-by-all, which is a different fastcore call rather than `target = query`. */
  target?: PointSet
  /** Neighbours per point for the tangent-vector fit. */
  k: number
  /** Point spacing in micrometres; `0` leaves the skeleton as traced. */
  resample: number
  normalize: boolean
  symmetry: NblastSymmetry
  useAlpha: boolean
}

export interface NblastResult {
  /** Row-major, `rows * cols`. */
  scores: Float64Array
  rows: number
  cols: number
}

/** Score one set against another, or a set against itself. */
export async function runNblast(
  request: NblastRequest,
  options: CallOptions = {},
): Promise<NblastResult> {
  const result = await callPython(
    { module: 'nblast', fn: 'coda_nblast_run', args: [request] },
    options,
  )

  const rows = numberFrom(result, 'rows')
  const cols = numberFrom(result, 'cols')
  const scores = float64From(result, 'scores')
  if (scores.length !== rows * cols) {
    throw new Error(
      `NBLAST returned ${scores.length} scores for a ${rows} x ${cols} comparison`,
    )
  }
  return { scores, rows, cols }
}

/**
 * The k nearest neighbours of each neuron, without the full matrix.
 *
 * The scores are exact NBLAST values; only which pairs were scored is approximate, which is
 * what makes this `n * nCandidates` rather than `n^2`. `symmetry` matters more here than for a
 * matrix and is applied *before* the top-k cut — once only k neighbours per row survive there
 * is no transpose left to symmetrise against.
 */
export type NblastKnnRequest = {
  query: PointSet
  /**
   * Absent means each neuron's neighbours are found among the others, and a neuron is excluded
   * from its own row. With a target set nothing is excluded, so a neuron present in both
   * matches itself at 1.0 — fastcore's behaviour, kept rather than corrected.
   */
  target?: PointSet
  /** Neighbours to return per neuron. */
  k: number
  /** Shortlist size per neuron: the one recall-against-cost knob. */
  nCandidates: number
  /** Neighbours per point for the tangent-vector fit — fastcore's other `k`. */
  tangentK: number
  resample: number
  normalize: boolean
  symmetry: NblastSymmetry
  useAlpha: boolean
}

export interface NblastKnnResult {
  /**
   * Row-major `rows * k` indices into the target set (or the query set for an all-by-all),
   * descending by score. A row with fewer than `k` candidates is padded with `-1`.
   */
  idx: Int32Array
  /** Aligned to `idx`; padding is `-Infinity`. */
  scores: Float64Array
  rows: number
  k: number
}

export async function runNblastKnn(
  request: NblastKnnRequest,
  options: CallOptions = {},
): Promise<NblastKnnResult> {
  const result = await callPython(
    { module: 'nblast', fn: 'coda_nblast_knn_run', args: [request] },
    options,
  )

  const rows = numberFrom(result, 'rows')
  const k = numberFrom(result, 'k')
  const idx = int32From(result, 'idx')
  const scores = float64From(result, 'scores')
  if (idx.length !== rows * k || scores.length !== rows * k) {
    throw new Error(`NBLAST k-NN returned ${idx.length} matches for ${rows} x ${k}`)
  }
  return { idx, scores, rows, k }
}

/**
 * One neuron set's synapses, flattened.
 *
 * `PointSet`'s sibling and deliberately not the same type: a dotprop set carries a tree, and
 * a synapse set carries a connector type instead. Neither field means anything to the other
 * call, and one type with both halves optional is how two callers come to disagree about
 * which of them is required.
 *
 * **Points are micrometres**, converted on the way in for the reason `PointSet` records.
 */
export type SynapseSet = {
  /** xyz interleaved, one neuron after the last. */
  points: Float32Array
  /**
   * A small integer per point — Coda maps `polarity` onto 0 for pre and 1 for post.
   *
   * Read only when `byType` is set. Where the point cloud carries no polarity column at all
   * every point is 0, which makes `byType` a comparison every synapse passes; the node turns
   * the control off rather than leaving it on and meaningless.
   */
  types: Int32Array
  /** Where each neuron starts, counted in points. Length is `count + 1`. */
  offsets: Int32Array
}

export type SynblastRequest = {
  query: SynapseSet
  /** Absent means an all-by-all, which is fastcore's own `target=None`. */
  target?: SynapseSet
  /** Compare a presynapse only against presynapses, and a postsynapse only against those. */
  byType: boolean
  normalize: boolean
  symmetry: NblastSymmetry
}

/**
 * Score two neuron sets by where their synapses are rather than by their shape.
 *
 * The result is a `NblastResult` and not a type of its own, because it is the same thing: a
 * row-major score matrix on the same scale, out of the same FCWB lookup table. What differs
 * is the question, and that belongs on the node rather than in the shape of the answer.
 */
export async function runSynblast(
  request: SynblastRequest,
  options: CallOptions = {},
): Promise<NblastResult> {
  const result = await callPython(
    { module: 'nblast', fn: 'coda_synblast_run', args: [request] },
    options,
  )

  const rows = numberFrom(result, 'rows')
  const cols = numberFrom(result, 'cols')
  const scores = float64From(result, 'scores')
  if (scores.length !== rows * cols) {
    throw new Error(
      `syNBLAST returned ${scores.length} scores for a ${rows} x ${cols} comparison`,
    )
  }
  return { scores, rows, cols }
}
